src/manifold/src/csg_tree.cpp

```diff
@@ -175,54 +181,66 @@ Manifold::Impl CsgLeafNode::Compose(
-              node->pImpl_->meshRelation_.triRef.end(),
-              combined.meshRelation_.triRef.begin() + nextTri,
-              UpdateMeshIDs({offset}));
+  // if we are already parallelizing for each node, do not perform multithreaded
+  // copying as it will slightly hurt performance
+  if (nodes.size() > 1 &&
+      (policy == ExecutionPolicy::Par ||
+       (policy == ExecutionPolicy::ParUnseq && !CudaEnabled())))
+    policy = ExecutionPolicy::Seq;
+
+  for_each_n_host(
+      nodes.size() > 1 ? ExecutionPolicy::Par : ExecutionPolicy::Seq,
+      countAt(0), nodes.size(),
+      [&nodes, &vertIndices, &edgeIndices, &triIndices, &combined,
+       policy](int i) {
+        auto &node = nodes[i];
+        if (node->transform_ == glm::mat4x3(1.0f)) {
+          copy(policy, node->pImpl_->vertPos_.begin(),
+               node->pImpl_->vertPos_.end(),
+               combined.vertPos_.begin() + vertIndices[i]);
+          copy(policy, node->pImpl_->faceNormal_.begin(),
+               node->pImpl_->faceNormal_.end(),
+               combined.faceNormal_.begin() + triIndices[i]);
+        } else {
+          // no need to apply the transform to the node, just copy the vertices
+          // and face normals and apply transform on the fly
+          auto vertPosBegin = thrust::make_transform_iterator(
+              node->pImpl_->vertPos_.begin(), Transform4x3({node->transform_}));
+          glm::mat3 normalTransform =
+              glm::inverse(glm::transpose(glm::mat3(node->transform_)));
+          auto faceNormalBegin = thrust::make_transform_iterator(
+              node->pImpl_->faceNormal_.begin(),
+              TransformNormals({normalTransform}));
+          copy_n(policy, vertPosBegin, node->pImpl_->vertPos_.size(),
+                 combined.vertPos_.begin() + vertIndices[i]);
+          copy_n(policy, faceNormalBegin, node->pImpl_->faceNormal_.size(),
+                 combined.faceNormal_.begin() + triIndices[i]);
+        }
+        copy(policy, node->pImpl_->halfedgeTangent_.begin(),
+             node->pImpl_->halfedgeTangent_.end(),
+             combined.halfedgeTangent_.begin() + edgeIndices[i]);
+        transform(
+            policy, node->pImpl_->halfedge_.begin(),
+            node->pImpl_->halfedge_.end(),
+            combined.halfedge_.begin() + edgeIndices[i],
+            UpdateHalfedge({vertIndices[i], edgeIndices[i], triIndices[i]}));
+        // Since the nodes may be copies containing the same meshIDs, it is
+        // important to add an offset so that each node instance gets
+        // unique meshIDs.
+        const int offset = i * Manifold::Impl::meshIDCounter_;
+        transform(policy, node->pImpl_->meshRelation_.triRef.begin(),
+                  node->pImpl_->meshRelation_.triRef.end(),
+                  combined.meshRelation_.triRef.begin() + triIndices[i],
+                  UpdateMeshIDs({offset}));
+      });
+
+  for (int i = 0; i < nodes.size(); i++) {
```

Comment:
@elalish this one

Reply:
Oh weird, seems like that should work, as long as the offsets are calculated up font like `vertIndices`. Mind giving it a shot here so I can see the test failures on the CI?

Reply:
done, note that it may not segfault but may loop indefinitely.

Reply:
Huh, it fails for parallel; I wonder if `std::map` doesn't play well with parallel insertions. Seems like that would be hard to manage...

Reply:
Oh ok, forgot that it is a map. This is normal then.

Reply:
removed that commit now